{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extraction of TCGA Metadata from Sevenbridges Genomics\n",
    "\n",
    "The purpose of this IPython notebook is to pull the majority of relevant TCGA metadata from a SPARQL ontology that is currently hosted by [Sevenbridges Genomics](https://cgc.sbgenomics.com/login/). By the end of this notebook, we will have stored them in the following data structures:\n",
    "\n",
    "1. A hierarchical dictionary of Pandas DataFrames, indexed by TCGA case (i.e. patient) universally unique identifiers (UUIDs)\n",
    "2. A hierarchical dictionary of Pandas DataFrames, indexed by TCGA file universally unique identifiers (UUIDs)\n",
    "3. One large, sorted CSV file that has been flattened to a 2D format but no longer has unique metadata fields (i.e. columns) due to mixed numbers of entries per patient.\n",
    "\n",
    "Note that the hierarchical dictionaries can be easily formatted for NoSQL databases, such as MongoDB, if desired. If you want more information on the ontology, it can be found by [clicking here](https://opensparql.sbgenomics.com/#/console). In addition to this notebook, we have written a Python script that pulls a subset of this metadata (~20 fields vs. 59 fields here) directly from your file directory on Sevenbridges Genomics website (first link above). A number of important considerations went into this analysis, and they are as follows:\n",
    "\n",
    "1. The SPARQL query had to be broken up into 2 separate queries, specifically two \"long\" queries. The practical reason for this is that a single POST-based retrieval was too large for the involved server to handle (likely due to the OPTIONAL query statements).\n",
    "2. Based on our desired analysis, we have implemented a variety of filters in the SPARQL query. These are given below and are the basis of the 2 queries. Note, if you want to run similar metadata query(ies) but have different filters, you should change these in **every** query below.\n",
    "\n",
    "    ```\n",
    "    # First filter by sample type\n",
    "    filter(?sample_type_label='Primary Tumor' || ?sample_type_label='Additional - New Primary' || ?sample_type_label='Additional Metastatic' || ?sample_type_label='Metastatic' || ?sample_type_label='Primary Blood Derived Cancer - Peripheral Blood' || ?sample_type_label='Recurrent Tumor' || ?sample_type_label='Blood Derived Normal' || ?sample_type_label='Bone Marrow Normal' || ?sample_type_label='Buccal Cell Normal' || ?sample_type_label='Solid Tissue Normal')\n",
    "\n",
    "    # Second filter by experimental strategy --> 'RNA-Seq' OR 'WGS'\n",
    "    filter(?experimental_strategy_label='RNA-Seq' || ?experimental_strategy_label='WGS') .\n",
    "\n",
    "    # Third filter by data type --> only 'Raw sequencing data'\n",
    "    filter(?data_type_label='Raw sequencing data') .\n",
    "\n",
    "    # Fourth filter by data format --> only 'BAM' files\n",
    "    filter(?data_format_label='BAM') .\n",
    "    ```\n",
    "\n",
    "3. One **major** consideration of this script (and likely the difficulty of this problem for most) is that the queried metadata is stored in a multiplexed, hierarchical, nested array structures (e.g. XML, JSON). Due to the variety of sampled fields, these data comprise a mix of static (e.g. demographic), dynamic (e.g. performance evaluations), and time-dependent (e.g. days until death) information that are not easily translatable to data formats most researchers are used to. For instance, take a moment to examine the table below. If one queries the metadata of a particular file, it is generally expected that there will be just _one_ Karnofsky performance score for the file in question. Yet, as shown below, one file apparently had 3 performance scores, another had none, and the last one had 2 scores diametrically opposite of one another (0 vs. 100). Without context, this result looks more like a corrupted database than a query problem, but it's actually the opposite. The issue actually arose because the query focused on file name and not the case UUID for the patient; in reality, the first file name belongs to a patient that received multiple performance scores across a duration of time. Thus, if all you selected for in your query was file name associated with patients' metadata, this is the type of result you would get.\n",
    "\n",
    "file_name | gdc_file_uuid | perf_score_karnofsky_label\n",
    "--- | --- | ---\n",
    "TCGA-KM-8442-10A-01D-2311-10_wgs_Illumina.bam |\t0273C55A-A89A-432E-A324-B90E1B28CED0 |\t90\n",
    "TCGA-KM-8442-10A-01D-2311-10_wgs_Illumina.bam |\t0273C55A-A89A-432E-A324-B90E1B28CED0 |\t100\n",
    "TCGA-KM-8442-10A-01D-2311-10_wgs_Illumina.bam |\t0273C55A-A89A-432E-A324-B90E1B28CED0 |\t80\n",
    "UNCID_1147822.fef221ff-634b-4188-92cf-5723b9a3c04d.sorted_genome_alignments.bam | 0273D20C-0DB7-44E0-B159-2A2D5FEB03D2 | Not available\n",
    "UNCID_1076554.a6af2ef5-6774-422c-99d1-826a686334d0.sorted_genome_alignments.bam |\t2FE88F0F-62BB-41BB-8B49-1B6E1E7EF368 |\t0\n",
    "UNCID_1076554.a6af2ef5-6774-422c-99d1-826a686334d0.sorted_genome_alignments.bam\t|\t2FE88F0F-62BB-41BB-8B49-1B6E1E7EF368 |\t100\n",
    "\n",
    "## The Takeaway\n",
    "Based on the above discussion, a decision was made to extract as much metadata as possible while retaining a minimum amount of the hierarchical structure. Through manipulations of Pandas data structures, this came about as a dictionary object with key-value pairs linked to their associated DataFrames. It should be noted that while they take a while to build (20-40 min), they permit the user to subsequently pull up all the metadata on a particular file or case (i.e. patient) by simply entering their file or case UUIDs as the dictionary key. Since these two key types (file UUID vs. case UUID) have different numbers of entries represented in the TCGA database and their associated dictionaries, they were created and packaged in separate Pickle binaries (i.e. Python dictionary files).\n",
    "\n",
    "For those who are uncomfortable with any hierarchical data manipulation, we have also produced a 2D table of _all_ the data herein (see very bottom of notebook). It is important to note that **none** of the metadata fields in this 2D table are entirely unique due to the metadata problems described above, meaning some level of grouping or summarization will need to be done by the end-user. It will also not be indexable by SQL-based formats, for those who use SQL-light or MySQL pipelines.\n",
    "\n",
    "Despite these limitations, we hope this notebook will be of use to you for future metadata analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Initialize\n",
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "from pandas.io.json import json_normalize\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json, csv, tqdm, pickle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Use the public endpoint\n",
    "\n",
    "sparql_endpoint = \"https://opensparql.sbgenomics.com/bigdata/namespace/tcga_metadata_kb/sparql\"\n",
    "\n",
    "# Initialize the SPARQL wrapper with the endpoint\n",
    "sparql = SPARQLWrapper(sparql_endpoint)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "longQuery1 = \"\"\"\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#> \n",
    "PREFIX tcga: <https://www.sbgenomics.com/ontologies/2014/11/tcga#> \n",
    " \n",
    "select distinct ?gdc_file_uuid ?file_name ?file_submitter_id ?case_name ?aliquot_name ?ref_genome_label ?sample_type_label ?experimental_strategy_label ?data_type_label ?data_format_label ?gender_label ?race_label ?ethnicity_label ?ageAtDiagnosis ?disease_type_label ?sample_label ?investigation_label ?histological_diagnosis_label ?primary_site_label ?prior_dx ?clinical_m_label ?clinical_n_label ?clinical_t_label ?pathologic_m_label ?pathologic_n_label ?pathologic_t_label ?pathologic_stage_label ?perf_score_karnofsky_label ?perf_score_eastern_cancer_oncology_group_label ?perf_score_timing_label ?primary_therapy_outcome_success_label ?vital_status_label ?new_tumor_event_label ?new_tumor_event_after_initial_trtmt ?radiation_therapy_code_label ?radiation_therapy_site_label ?radiation_therapy_type_label ?days_to_last_followup ?year_of_diagnosis ?icd10 ?icd03_histology_label ?icd03_histology_site ?data_submitting_center_label ?seq_platform_label ?aliquot_concentration  ?analyte_A260A280Ratio ?analyte_type_label ?analyte_amount ?analyte_well_number ?spectrophotometer_method_label ?file_upload_date ?file_published_date ?file_last_modified_date ?portion_weight ?portion_is_ffpe ?portion_number ?portion_slide_label ?freezing_method_label ?tissue_source_site_label ?country_of_sample_procurement\n",
    "{ \n",
    "  ?file a tcga:File .\n",
    "  ?file rdfs:label ?file_name .\n",
    "  \n",
    "  ?file tcga:hasGDCFileUUID ?gdc_file_uuid .\n",
    "  \n",
    "  ?file tcga:hasSample ?sample .\n",
    "  ?sample tcga:hasSampleType ?st .\n",
    "  ?st rdfs:label ?sample_type_label \n",
    "  filter(?sample_type_label='Primary Tumor' || ?sample_type_label='Additional - New Primary' || ?sample_type_label='Additional Metastatic' || ?sample_type_label='Metastatic' || ?sample_type_label='Primary Blood Derived Cancer - Peripheral Blood' || ?sample_type_label='Recurrent Tumor' || ?sample_type_label='Blood Derived Normal' || ?sample_type_label='Bone Marrow Normal' || ?sample_type_label='Buccal Cell Normal' || ?sample_type_label='Solid Tissue Normal')\n",
    "      \n",
    "  ?file tcga:hasExperimentalStrategy ?xs .\n",
    "  ?xs rdfs:label ?experimental_strategy_label .\n",
    "  filter(?experimental_strategy_label='RNA-Seq' || ?experimental_strategy_label='WGS') .\n",
    "  \n",
    "  ?file tcga:hasDataType ?type .\n",
    "  ?type rdfs:label ?data_type_label .\n",
    "  filter(?data_type_label='Raw sequencing data') .\n",
    "  \n",
    "  ?file tcga:hasDataFormat ?format .\n",
    "  ?format rdfs:label ?data_format_label .\n",
    "  filter(?data_format_label='BAM') .\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasSubmitterId ?file_submitter_id .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPerformanceStatusScoreKarnofsky ?perf_score_karnofsky .\n",
    "  ?perf_score_karnofsky rdfs:label ?perf_score_karnofsky_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .\n",
    "  ?case rdfs:label ?case_name .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAliquot ?aliquot .  \n",
    "  ?aliquot rdfs:label ?aliquot_name .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasReferenceGenome ?ref_genome .   \n",
    "  ?ref_genome rdfs:label ?ref_genome_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .\n",
    "  ?case tcga:hasGender ?gender .\n",
    "  ?gender rdfs:label ?gender_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasEthnicity ?ethnicity .\n",
    "  ?ethnicity rdfs:label ?ethnicity_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasRace ?race .\n",
    "  ?race rdfs:label ?race_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasAgeAtDiagnosis ?ageAtDiagnosis .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasDiseaseType ?diseaseType .\n",
    "  ?diseaseType rdfs:label ?disease_type_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasSample ?sample .\n",
    "  ?sample rdfs:label ?sample_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .  \n",
    "  ?case tcga:hasInvestigation ?investigation .\n",
    "  ?investigation rdfs:label ?inv_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case . \n",
    "  ?case tcga:hasHistologicalDiagnosis ?hd .\n",
    "  ?hd rdfs:label ?histological_diagnosis_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPrimarySite ?primary_site .\n",
    "  ?primary_site rdfs:label ?primary_site_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPriorDiagnosis ?prior_dx_base .\n",
    "  ?prior_dx_base rdfs:label ?prior_dx .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasClinicalM ?clinical_m .\n",
    "  ?clinical_m rdfs:label ?clinical_m_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasClinicalM ?clinical_n .\n",
    "  ?clinical_n rdfs:label ?clinical_n_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasClinicalM ?clinical_t .\n",
    "  ?clinical_t rdfs:label ?clinical_t_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasClinicalStage ?clinical_stage .\n",
    "  ?clinical_stage rdfs:label ?clinical_stage_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPathologicM ?pathologic_m .\n",
    "  ?pathologic_m rdfs:label ?pathologic_m_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPathologicN ?pathologic_n .\n",
    "  ?pathologic_n rdfs:label ?pathologic_n_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPathologicT ?pathologic_t .\n",
    "  ?pathologic_t rdfs:label ?pathologic_t_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPathologicStage ?pathologic_stage .\n",
    "  ?pathologic_stage rdfs:label ?pathologic_stage_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPerformanceStatusScoreECOG ?perf_score_eastern_cancer_oncology_group .\n",
    "  ?perf_score_eastern_cancer_oncology_group rdfs:label ?perf_score_eastern_cancer_oncology_group_label .\n",
    "  }\n",
    "}\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Set SPARQL extraction method to POST, return in the JSON format, and then convert JSON to a Python dictionary object\n",
    "sparql.method = 'POST'\n",
    "sparql.setReturnFormat(JSON)\n",
    "sparql.setQuery(longQuery1)\n",
    "resultsLong1 = sparql.query().convert()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "longQuery2 = \"\"\"\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#> \n",
    "PREFIX tcga: <https://www.sbgenomics.com/ontologies/2014/11/tcga#> \n",
    " \n",
    "select distinct ?gdc_file_uuid ?file_name ?file_submitter_id ?case_name ?aliquot_name ?ref_genome_label ?sample_type_label ?experimental_strategy_label ?data_type_label ?data_format_label ?gender_label ?race_label ?ethnicity_label ?ageAtDiagnosis ?disease_type_label ?sample_label ?investigation_label ?histological_diagnosis_label ?primary_site_label ?prior_dx ?clinical_m_label ?clinical_n_label ?clinical_t_label ?pathologic_m_label ?pathologic_n_label ?pathologic_t_label ?pathologic_stage_label ?perf_score_karnofsky_label ?perf_score_eastern_cancer_oncology_group_label ?perf_score_timing_label ?primary_therapy_outcome_success_label ?vital_status_label ?new_tumor_event_label ?new_tumor_event_after_initial_trtmt ?radiation_therapy_code_label ?radiation_therapy_site_label ?radiation_therapy_type_label ?days_to_last_followup ?year_of_diagnosis ?icd10 ?icd03_histology_label ?icd03_histology_site ?data_submitting_center_label ?seq_platform_label ?aliquot_concentration  ?analyte_A260A280Ratio ?analyte_type_label ?analyte_amount ?analyte_well_number ?spectrophotometer_method_label ?file_upload_date ?file_published_date ?file_last_modified_date ?portion_weight ?portion_is_ffpe ?portion_number ?portion_slide_label ?freezing_method_label ?tissue_source_site_label ?country_of_sample_procurement\n",
    "{ \n",
    "  ?file a tcga:File .\n",
    "  ?file rdfs:label ?file_name .\n",
    "  \n",
    "  ?file tcga:hasGDCFileUUID ?gdc_file_uuid .\n",
    "  \n",
    "  ?file tcga:hasSample ?sample .\n",
    "  ?sample tcga:hasSampleType ?st .\n",
    "  ?st rdfs:label ?sample_type_label \n",
    "  filter(?sample_type_label='Primary Tumor' || ?sample_type_label='Additional - New Primary' || ?sample_type_label='Additional Metastatic' || ?sample_type_label='Metastatic' || ?sample_type_label='Primary Blood Derived Cancer - Peripheral Blood' || ?sample_type_label='Recurrent Tumor' || ?sample_type_label='Blood Derived Normal' || ?sample_type_label='Bone Marrow Normal' || ?sample_type_label='Buccal Cell Normal' || ?sample_type_label='Solid Tissue Normal')\n",
    "      \n",
    "  ?file tcga:hasExperimentalStrategy ?xs .\n",
    "  ?xs rdfs:label ?experimental_strategy_label .\n",
    "  filter(?experimental_strategy_label='RNA-Seq' || ?experimental_strategy_label='WGS') .\n",
    "  \n",
    "  ?file tcga:hasDataType ?type .\n",
    "  ?type rdfs:label ?data_type_label .\n",
    "  filter(?data_type_label='Raw sequencing data') .\n",
    "  \n",
    "  ?file tcga:hasDataFormat ?format .\n",
    "  ?format rdfs:label ?data_format_label .\n",
    "  filter(?data_format_label='BAM') .\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPerformanceStatusScoreTiming ?perf_score_timing .\n",
    "  ?perf_score_timing rdfs:label ?perf_score_timing_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasPrimaryTherapyOutcomeSuccess ?primary_therapy_outcome_success .\n",
    "  ?primary_therapy_outcome_success rdfs:label ?primary_therapy_outcome_success_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasVitalStatus ?vital_status .\n",
    "  ?vital_status rdfs:label ?vital_status_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasNewTumorEvent ?new_tumor_event .\n",
    "  ?new_tumor_event rdfs:label ?new_tumor_event_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasRadiationTherapy ?radiation_therapy .\n",
    "  ?radiation_therapy rdfs:label ?radiation_therapy_code_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasRadiationTherapy ?radiation_therapy .\n",
    "  ?radiation_therapy tcga:hasRadiationTherapySite ?radiation_therapy_site .\n",
    "  ?radiation_therapy_site rdfs:label ?radiation_therapy_site_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasRadiationTherapy ?radiation_therapy .\n",
    "  ?radiation_therapy tcga:hasRadiationType ?radiation_therapy_type .\n",
    "  ?radiation_therapy_type rdfs:label ?radiation_therapy_type_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasNewTumorEventAfterInitialTreatment ?new_tumor_event_after_initial_trtmt .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasDaysToLastFollowUp ?days_to_last_followup .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasYearOfDiagnosis ?year_of_diagnosis .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .   \n",
    "  ?case tcga:hasIcd10 ?icd10 .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case . \n",
    "  ?case tcga:hasIcdO3Histology ?icd03_histology_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case . \n",
    "  ?case tcga:hasIcdO3Site ?icd03_histology_site .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasDataSubmittingCenter ?data_submitting_center .   \n",
    "  ?data_submitting_center rdfs:label ?data_submitting_center_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasPlatform ?seq_platform .   \n",
    "  ?seq_platform rdfs:label ?seq_platform_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAliquot ?aliquot .   \n",
    "  ?aliquot tcga:hasConcentration ?aliquot_concentration .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .   \n",
    "  ?analyte tcga:hasA260A280Ratio ?analyte_A260A280Ratio .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .   \n",
    "  ?analyte tcga:hasAnalyteType ?analyte_type .\n",
    "  ?analyte_type rdfs:label ?analyte_type_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .   \n",
    "  ?analyte tcga:hasAmount ?analyte_amount .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .   \n",
    "  ?analyte rdfs:label ?analyte_well_number .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .   \n",
    "  ?analyte tcga:hasSpectrophotometerMethod ?spectrophotometer_method .\n",
    "  ?spectrophotometer_method rdfs:label ?spectrophotometer_method_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:uploadDate ?file_upload_date .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:publishedDate ?file_published_date .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:lastModifiedDate ?file_last_modified_date .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .\n",
    "  ?analyte tcga:hasPortion ?portion .\n",
    "  ?portion tcga:hasWeight ?portion_weight .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .\n",
    "  ?analyte tcga:hasPortion ?portion .\n",
    "  ?portion tcga:hasIsFFPE ?portion_is_ffpe .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .\n",
    "  ?analyte tcga:hasPortion ?portion .\n",
    "  ?portion tcga:hasPortionNumber ?portion_number .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasAnalyte ?analyte .\n",
    "  ?analyte tcga:hasPortion ?portion .\n",
    "  ?portion tcga:hasSlide ?portion_slide .\n",
    "  ?portion_slide rdfs:label ?portion_slide_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .\n",
    "  ?case tcga:hasSample ?sample .\n",
    "  ?sample tcga:hasFreezingMethod ?freezing_method .\n",
    "  ?freezing_method rdfs:label ?freezing_method_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .\n",
    "  ?case tcga:hasSample ?sample .\n",
    "  ?sample tcga:hasTissueSourceSite ?tissue_source_site .\n",
    "  ?tissue_source_site rdfs:label ?tissue_source_site_label .\n",
    "  }\n",
    "  \n",
    "  OPTIONAL {\n",
    "  ?file tcga:hasCase ?case .\n",
    "  ?case tcga:hasSample ?sample .\n",
    "  ?sample tcga:hasCountryOfSampleProcurement ?country_of_sample_procurement .\n",
    "  }\n",
    "}\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Set SPARQL extraction method to POST, return in the JSON format, and then convert JSON to a Python dictionary object\n",
    "sparql.method = 'POST'\n",
    "sparql.setReturnFormat(JSON)\n",
    "sparql.setQuery(longQuery2)\n",
    "resultsLong2 = sparql.query().convert()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of long query 1 list is 18669\n",
      "Length of long query 2 list is 31766\n"
     ]
    }
   ],
   "source": [
    "# Find the length of the \"long\" queries\n",
    "# Note that the extended lengths are likely due to the duplicate entries discussed at the beginning of this notebook\n",
    "print(\"Length of long query 1 list is %r\" %len(resultsLong1['results']['bindings']))\n",
    "print(\"Length of long query 2 list is %r\" %len(resultsLong2['results']['bindings']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 18669/18669 [01:19<00:00, 236.11it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18669, 28)\n"
     ]
    }
   ],
   "source": [
    "# Convert individual Python dictionaries into Pandas DataFrame objects, and then concatenate the result\n",
    "# This is for \"long\" query 1\n",
    "framesListLong1 = []\n",
    "for ind in tqdm.trange(0,len(resultsLong1['results']['bindings'])):\n",
    "    framesListLong1.append(pd.DataFrame.from_dict(resultsLong1['results']['bindings'][ind]))\n",
    "\n",
    "# Note that the dictionaries are nested and that the data we want are stored with the 'value' key.\n",
    "# The issue is that sub-dictionaries (i.e. in a key:value format) is that the their values cannot be extracted directly.\n",
    "# We solve this by concatenating all of the nested keys (e.g. 'type', 'value') as rows in the DataFrame objects and \n",
    "# then slicing the resultant object based on the index key 'value'. One should note that this does not yield a unique\n",
    "# index, but this is fixed via the join procedures later on before the CSV file is written.\n",
    "concatDFLong1 = pd.concat(framesListLong1).loc['value',:]\n",
    "print(concatDFLong1.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 31766/31766 [01:43<00:00, 306.39it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(31766, 37)\n"
     ]
    }
   ],
   "source": [
    "# Convert individual Python dictionaries into Pandas DataFrame objects, and then concatenate the result\n",
    "# This is for \"long\" query 2\n",
    "framesListLong2 = []\n",
    "for ind in tqdm.trange(0,len(resultsLong2['results']['bindings'])):\n",
    "    framesListLong2.append(pd.DataFrame.from_dict(resultsLong2['results']['bindings'][ind]))\n",
    "    \n",
    "# As described before, note that the dictionaries are nested and that the data we want are stored with the 'value' key.\n",
    "# The issue is that sub-dictionaries (i.e. in a key:value format) is that the their values cannot be extracted directly.\n",
    "# We solve this by concatenating all of the nested keys (e.g. 'type', 'value') as rows in the DataFrame objects and \n",
    "# then slicing the resultant object based on the index key 'value'. One should note that this does not yield a unique\n",
    "# index, but this is fixed via the join procedures later on before the CSV file is written.\n",
    "concatDFLong2 = pd.concat(framesListLong2).loc['value',:]\n",
    "print(concatDFLong2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18669, 28)\n",
      "(31766, 32)\n",
      "(33263, 59)\n",
      "The number of unique files are: 18154 (should be 18154)\n",
      "The number of unique cases are: 10499 (should be 10499)\n"
     ]
    }
   ],
   "source": [
    "# Now, we're going to create the equivalent of a 3D lookup table\n",
    "# using Pandas DataFrames and a Python dictionary.\n",
    "# First, drop all duplicate columns except 'gdc_file\"uuid' on the 'right-hand' file and \n",
    "# then do a left join on the results of the two long queries:\n",
    "\n",
    "reformattedConcatDFLong2 = concatDFLong2.drop(['file_name','data_format_label','data_type_label', 'experimental_strategy_label', 'sample_type_label'], axis=1)\n",
    "\n",
    "print(concatDFLong1.shape)\n",
    "print(formattedConcatDFLong2.shape)\n",
    "\n",
    "# Note that 'gdc_file_uuid' is the only, truly unique identifier in this dataset\n",
    "combinedDFLong = pd.merge(concatDFLong1, reformattedConcatDFLong2, on = 'gdc_file_uuid')\n",
    "\n",
    "# Check the resulting DataFrame dimensions and number of unique files\n",
    "print(combinedDFLong.shape)\n",
    "print(\"The number of unique files are: %r (should be 18154)\" %len(combinedDFLong['gdc_file_uuid'].unique()))\n",
    "print(\"The number of unique cases are: %r (should be 10499)\" %len(combinedDFLong['case_name'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Next, create groupby objects, one by 'gdc_file_uuid' and one by 'case_name'\n",
    "fileGroupedDF = combinedDFLong.groupby('gdc_file_uuid')\n",
    "caseGroupedDF = combinedDFLong.groupby('case_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10499/10499 [15:52<00:00, 13.11it/s]\n"
     ]
    }
   ],
   "source": [
    "# Lastly, create dictionaries of DataFrames for each 'gdc_file_uuid' and 'case_name'.\n",
    "# The dictionary for 'case_name' is created in this cell and 'gdc_file_uuid' in the next cell.\n",
    "# Takes a while to run: est. 18 min on Mac OS X\n",
    "# with 2.9 GHz dual-core Intel i7 and 8 GB DDR3 RAM.\n",
    "caseGroupedDict = {}\n",
    "for v in tqdm.tqdm(list(caseGroupedDF.groups.keys())):\n",
    "    caseGroupedDict.update({v: combinedDFLong.groupby(['case_name']).get_group(v)})\n",
    "#     print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 18154/18154 [46:19<00:00,  8.15it/s]\n"
     ]
    }
   ],
   "source": [
    "# Takes a while to run: est. 50 min on Mac OS X\n",
    "# with 2.9 GHz dual-core Intel i7 and 8 GB DDR3 RAM.\n",
    "fileGroupedDict = {}\n",
    "for v in tqdm.tqdm(list(fileGroupedDF.groups.keys())):\n",
    "    fileGroupedDict.update({v: combinedDFLong.groupby(['gdc_file_uuid']).get_group(v)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Save dict/DataFrame as a pickle file\n",
    "pickle.dump(caseGroupedDict, open(\"caseGroupedDict.p\", \"wb\"), protocol=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Save dict/DataFrame as a pickle file\n",
    "pickle.dump(fileGroupedDict, open(\"fileGroupedDict.p\", \"wb\"), protocol=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load pickle files to demonstrate that the Pandas formatting is not lost in the binary conversions\n",
    "caseTesting = pickle.load(open(\"caseGroupedDict.p\", \"rb\"))\n",
    "fileNameTesting = pickle.load(open(\"fileGroupedDict.p\", \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ageAtDiagnosis</th>\n",
       "      <th>aliquot_name</th>\n",
       "      <th>case_name</th>\n",
       "      <th>clinical_m_label</th>\n",
       "      <th>clinical_n_label</th>\n",
       "      <th>clinical_t_label</th>\n",
       "      <th>data_format_label</th>\n",
       "      <th>data_type_label</th>\n",
       "      <th>disease_type_label</th>\n",
       "      <th>ethnicity_label</th>\n",
       "      <th>...</th>\n",
       "      <th>portion_weight</th>\n",
       "      <th>primary_therapy_outcome_success_label</th>\n",
       "      <th>radiation_therapy_code_label</th>\n",
       "      <th>radiation_therapy_site_label</th>\n",
       "      <th>radiation_therapy_type_label</th>\n",
       "      <th>seq_platform_label</th>\n",
       "      <th>spectrophotometer_method_label</th>\n",
       "      <th>tissue_source_site_label</th>\n",
       "      <th>vital_status_label</th>\n",
       "      <th>year_of_diagnosis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16759</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16760</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ageAtDiagnosis aliquot_name case_name clinical_m_label clinical_n_label  \\\n",
       "16759           True         True      True             True             True   \n",
       "16760           True         True      True             True             True   \n",
       "\n",
       "      clinical_t_label data_format_label data_type_label disease_type_label  \\\n",
       "16759             True              True            True               True   \n",
       "16760             True              True            True               True   \n",
       "\n",
       "      ethnicity_label        ...        portion_weight  \\\n",
       "16759            True        ...                  True   \n",
       "16760            True        ...                  True   \n",
       "\n",
       "      primary_therapy_outcome_success_label radiation_therapy_code_label  \\\n",
       "16759                                  True                        False   \n",
       "16760                                  True                        False   \n",
       "\n",
       "      radiation_therapy_site_label radiation_therapy_type_label  \\\n",
       "16759                        False                        False   \n",
       "16760                        False                        False   \n",
       "\n",
       "      seq_platform_label spectrophotometer_method_label  \\\n",
       "16759               True                           True   \n",
       "16760               True                           True   \n",
       "\n",
       "      tissue_source_site_label vital_status_label year_of_diagnosis  \n",
       "16759                     True               True              True  \n",
       "16760                     True               True              True  \n",
       "\n",
       "[2 rows x 59 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test for equality between saved and loaded file as a sanity check.\n",
    "# Note that values saved as 'NaN' will show as 'False'.\n",
    "caseTestingKey = '40C217EE-429F-41F4-ADC6-DB9186622B17'\n",
    "caseTesting[caseTestingKey] == caseGroupedDict[caseTestingKey]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ageAtDiagnosis</th>\n",
       "      <th>aliquot_name</th>\n",
       "      <th>case_name</th>\n",
       "      <th>clinical_m_label</th>\n",
       "      <th>clinical_n_label</th>\n",
       "      <th>clinical_t_label</th>\n",
       "      <th>data_format_label</th>\n",
       "      <th>data_type_label</th>\n",
       "      <th>disease_type_label</th>\n",
       "      <th>ethnicity_label</th>\n",
       "      <th>...</th>\n",
       "      <th>portion_weight</th>\n",
       "      <th>primary_therapy_outcome_success_label</th>\n",
       "      <th>radiation_therapy_code_label</th>\n",
       "      <th>radiation_therapy_site_label</th>\n",
       "      <th>radiation_therapy_type_label</th>\n",
       "      <th>seq_platform_label</th>\n",
       "      <th>spectrophotometer_method_label</th>\n",
       "      <th>tissue_source_site_label</th>\n",
       "      <th>vital_status_label</th>\n",
       "      <th>year_of_diagnosis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5070</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ageAtDiagnosis aliquot_name case_name clinical_m_label clinical_n_label  \\\n",
       "5070           True         True      True             True             True   \n",
       "\n",
       "     clinical_t_label data_format_label data_type_label disease_type_label  \\\n",
       "5070             True              True            True               True   \n",
       "\n",
       "     ethnicity_label        ...        portion_weight  \\\n",
       "5070            True        ...                  True   \n",
       "\n",
       "     primary_therapy_outcome_success_label radiation_therapy_code_label  \\\n",
       "5070                                  True                        False   \n",
       "\n",
       "     radiation_therapy_site_label radiation_therapy_type_label  \\\n",
       "5070                        False                        False   \n",
       "\n",
       "     seq_platform_label spectrophotometer_method_label  \\\n",
       "5070               True                           True   \n",
       "\n",
       "     tissue_source_site_label vital_status_label year_of_diagnosis  \n",
       "5070                     True               True              True  \n",
       "\n",
       "[1 rows x 59 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test for equality between saved and loaded file as a sanity check.\n",
    "# Note that values saved as 'NaN' will show as 'False'.\n",
    "fileTestingKey = '01249B8C-2E9E-4CEA-B39A-485863AB231A'\n",
    "fileNameTesting[fileTestingKey] == fileGroupedDict[fileTestingKey]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Lastly, we're going to make one large metadata table. Some prefer 2D tables versus\n",
    "# hierarchical data, so this is for those folks. Do note, however, that no metadata field will\n",
    "# have unique values. We will order the table based first on 'case_name', secondly on 'gdc_file_uuid',\n",
    "# and then we'll sort the rest of the metadata fields alphabetically to try to help with the table's readability.\n",
    "\n",
    "sortedCombinedDFLong = combinedDFLong.sort_values(by = ['case_name', 'gdc_file_uuid'])\n",
    "cols = sortedCombinedDFLong.columns.tolist()\n",
    "sortedCols = sorted(cols)\n",
    "sortedCols.remove('case_name')\n",
    "sortedCols.remove('gdc_file_uuid')\n",
    "sortedCols.insert(0,'gdc_file_uuid')\n",
    "sortedCols.insert(0,'case_name')\n",
    "alphaSortedCombinedDFLong = sortedCombinedDFLong[sortedCols]\n",
    "\n",
    "# Drop the previous index (rearranged due to sorting) and reset to a new integer index\n",
    "alphaSortedCombinedDFLongNewIndex = alphaSortedCombinedDFLong.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write the DataFrame as a CSV file to your local directory\n",
    "alphaSortedCombinedDFLongNewIndex.to_csv('mergedCGCMetadata_tcga_microbiome.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {
   "attach-environment": true,
   "environment": "py3k",
   "summary": "cgc_sparql_3D_dict_arrays",
   "url": "https://anaconda.org/gregpoore/cgc_sparql_3d_dict_arrays"
  },
  "kernelspec": {
   "display_name": "Python [py3k]",
   "language": "python",
   "name": "Python [py3k]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
